File: modules/core/src/test/clojure/org/immutant/core/test/ClojureMetaData.clj
;; Copyright 2008-2012 Red Hat, Inc, and individual contributors.
;; 
;; This is free software; you can redistribute it and/or modify it
;; under the terms of the GNU Lesser General Public License as
;; published by the Free Software Foundation; either version 2.1 of
;; the License, or (at your option) any later version.
;; 
;; This software is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
;; Lesser General Public License for more details.
;; 
;; You should have received a copy of the GNU Lesser General Public
;; License along with this software; if not, write to the Free
;; Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
;; 02110-1301 USA, or see the FSF site: http://www.fsf.org.

(ns org.immutant.core.test.ClojureMetaData
  (:use clojure.test)
  (:import [org.immutant.core ClojureMetaData])
  (:require [clojure.java.io            :as io]
            [immutant.runtime.bootstrap :as bootstrap]))

(let [descriptor (ClojureMetaData/parse (io/file (io/resource "simple-descriptor.clj")))
      md (ClojureMetaData. "app-name" descriptor)]
  (deftest parse "it should parse the descriptor and return a map"
    (is (= "my.namespace/init" (.get descriptor "init"))))

  (deftest getString "it should return the proper value as a String"
    (let [value (.getString md "init")]
      (is (= "my.namespace/init" value))
      (is (instance? String value))))

  (deftest getInitFunction "it should return the proper value"
    (is (= "my.namespace/init" (.getInitFunction md))))

  (deftest it-should-allow-access-to-any-metadata-value
    (is (= "biscuit" (.getString md "ham"))))

  (let [merged-md (doto (ClojureMetaData. "app-name" descriptor)
                    (.setConfig (bootstrap/read-and-stringify-full-app-config
                                 (io/file (io/resource "simple-descriptor.clj"))
                                 (io/file (io/resource "project-root")) )))]
    (deftest the-immutant-map-from-project-clj-should-be-included
      (is (= "gravy" (.getString merged-md "biscuit"))))

    (deftest the-immutant-map-from-project-clj-should-take-a-backseat-to-existing-config
      (is (= "biscuit" (.getString merged-md "ham"))))

    (deftest the-default-resolve-dependencies-for-non-archives-should-be-true
      (is (= true (.resolveDependencies md))))

    (deftest the-default-resolve-dependencies-for-archives-should-also-be-true
      (.explode md (io/file ""))
      (is (= true (.resolveDependencies md))))
    
    (deftest a-resolve-dependencies-value-in-project-clj-should-win
      (is (= false (.resolveDependencies merged-md))))))

(let [md (ClojureMetaData. "app-name"
                           (ClojureMetaData/parse (io/file (io/resource "hashy-descriptor.clj"))))]
  
  (deftest getHash "it should return the proper value as a Hash"
    (is (= {"biscuit" "gravy"} (.getHash md "ham"))))
  
  (deftest getList "it should return the proper value as a List"
    (is (= [:biscuit :gravy] (.getList md "foo"))))

  (deftest getLeinProfiles "it should stringify the values so we can share across runtimes"
    (is (= [":dev" ":prod"] (.getLeinProfiles md)))))

File: modules/core/src/main/java/org/immutant/core/ClojureMetaData.java
/*
 * Copyright 2008-2013 Red Hat, Inc, and individual contributors.
 * 
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

package org.immutant.core;

import java.io.File;
import java.util.List;
import java.util.Map;

import org.immutant.bootstrap.ApplicationBootstrapUtils;
import org.jboss.as.server.deployment.AttachmentKey;
import org.jboss.as.server.deployment.DeploymentUnit;
import org.projectodd.polyglot.core.app.ApplicationMetaData;

public class ClojureMetaData extends ApplicationMetaData {

    public static final AttachmentKey<ClojureMetaData> ATTACHMENT_KEY = AttachmentKey.create( ClojureMetaData.class );
    public static final AttachmentKey<File> DESCRIPTOR_FILE = AttachmentKey.create( File.class );
    public static final AttachmentKey<String> FULL_APP_CONFIG = AttachmentKey.create( String.class );
    public static final AttachmentKey<String> LEIN_PROJECT = AttachmentKey.create( String.class );

    public ClojureMetaData(String applicationName, Map<String, Object> config) {
        super( applicationName ); 
        setConfig( config );
        String root = getString( "root" );
        if (root != null) {
            setRoot( new File( root ) );
        }
    }

    @Override
    public void attachTo(DeploymentUnit unit) {
        super.attachTo( unit );
        unit.putAttachment( ATTACHMENT_KEY, this );
    }

    public String getInitFunction() {
        return getString( "init" );
    }

    @SuppressWarnings("rawtypes")
    public List getLeinProfiles() {
        return getList( "lein-profiles" );
    }
    
    /**
     * See if the user has explicitly set :resolve-dependencies. If not, 
     * resolve by default.
     */
    public boolean resolveDependencies() {
        if (this.config.containsKey( "resolve-dependencies" )) {
            return (Boolean)get( "resolve-dependencies" );
        } else {
            return true;
        }
    }
    
    public String getString(String key) {
        return (String)get( key );
    }

    @SuppressWarnings("unchecked")
    public Map<String, ?> getHash(String key) {
        return (Map<String, Object>)get( key );
    }

    @SuppressWarnings("rawtypes")
    public List getList(String key) {
        return (List) get( key );
    }

    public Object get(String key) {
        return this.config.get( key );
    }
    
    @SuppressWarnings({ "rawtypes", "unchecked" })
    public void setConfig(Map config) {
        this.config = config;
    }

    public Map<String, Object> getConfig() {
        return this.config;
    }

    @SuppressWarnings("unchecked")
    public static Map<String, Object> parse(File file) throws Exception {
        return ApplicationBootstrapUtils.parseDescriptor( file );
    }

    private Map<String, Object> config;
}
